File: Sources/AdbShell/Services/DeviceStatsParser.swift
import Foundation

/// Разбор вывода трёх shell-команд, из которых собирается `DeviceStats`:
/// `dumpsys cpuinfo`, `cat /proc/meminfo` и `dumpsys battery`.
enum DeviceStatsParser {
    /// Строка "TOTAL" в `dumpsys cpuinfo` выглядит как "23% TOTAL: 12% user + ..."
    /// — берём число перед первым "% TOTAL".
    static func parseCpuPercent(_ output: String) -> Double? {
        for line in output.split(separator: "\n") {
            guard let range = line.range(of: "% TOTAL") else { continue }
            let prefix = line[line.startIndex..<range.lowerBound]
            let digits = prefix.trimmingCharacters(in: .whitespaces)
                .split(separator: " ").last.map(String.init) ?? String(prefix)
            if let value = Double(digits.trimmingCharacters(in: .whitespaces)) {
                return min(max(value, 0), 100)
            }
        }
        return nil
    }

    /// `/proc/meminfo`: используем MemTotal - MemAvailable (либо MemFree, если
    /// MemAvailable недоступен — на старых ядрах его может не быть).
    static func parseMemInfo(_ output: String) -> (usedKB: Int, totalKB: Int)? {
        var total: Int?
        var available: Int?
        var free: Int?
        for line in output.split(separator: "\n") {
            let parts = line.split(separator: ":")
            guard parts.count == 2 else { continue }
            let key = parts[0].trimmingCharacters(in: .whitespaces)
            let valueDigits = parts[1].trimmingCharacters(in: .whitespaces)
                .split(separator: " ").first.map(String.init)
            guard let valueDigits, let value = Int(valueDigits) else { continue }
            switch key {
            case "MemTotal": total = value
            case "MemAvailable": available = value
            case "MemFree": free = value
            default: break
            }
        }
        guard let total else { return nil }
        let used = total - (available ?? free ?? total)
        return (usedKB: max(used, 0), totalKB: total)
    }

    /// `dumpsys battery`: level/scale дают процент, temperature — в десятых долях °C,
    /// заряд определяем по status==2 (BATTERY_STATUS_CHARGING) либо по *_powered.
    static func parseBattery(_ output: String) -> (level: Int?, temperature: Double?, charging: Bool) {
        var level: Int?
        var scale: Int?
        var temperature: Double?
        var status: Int?
        var anyPowered = false

        for rawLine in output.split(separator: "\n") {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard let colon = line.firstIndex(of: ":") else { continue }
            let key = line[line.startIndex..<colon].trimmingCharacters(in: .whitespaces)
            let value = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            switch key {
            case "level": level = Int(value)
            case "scale": scale = Int(value)
            case "status": status = Int(value)
            case "temperature": temperature = Double(value).map { $0 / 10 }
            case "AC powered", "USB powered", "Wireless powered", "Dock powered":
                if value == "true" { anyPowered = true }
            default: break
            }
        }

        let percent = level.map { lvl -> Int in
            guard let scale, scale > 0, scale != 100 else { return lvl }
            return Int((Double(lvl) / Double(scale) * 100).rounded())
        }
        let charging = status == 2 || anyPowered
        return (level: percent, temperature: temperature, charging: charging)
    }

    static func parse(cpuOutput: String, memOutput: String, batteryOutput: String, timestamp: Date = Date()) -> DeviceStats {
        let mem = parseMemInfo(memOutput)
        let battery = parseBattery(batteryOutput)
        return DeviceStats(
            cpuPercent: parseCpuPercent(cpuOutput),
            memUsedKB: mem?.usedKB ?? 0,
            memTotalKB: mem?.totalKB ?? 0,
            batteryLevel: battery.level,
            batteryTemperature: battery.temperature,
            isCharging: battery.charging,
            timestamp: timestamp
        )
    }
}

File: desktop/src/main/adb/parsers/DeviceStatsParser.ts
// Порт Sources/AdbShell/Services/DeviceStatsParser.swift — разбор трёх
// shell-команд, из которых собирается DeviceStats: `dumpsys cpuinfo`,
// `cat /proc/meminfo`, `dumpsys battery`.

import { DeviceStats } from '../types/DeviceStats';

/** Строка "TOTAL" в `dumpsys cpuinfo` выглядит как "23% TOTAL: 12% user + ..."
 * — берём число перед первым "% TOTAL". */
export function parseCpuPercent(output: string): number | undefined {
  for (const line of output.split('\n')) {
    const idx = line.indexOf('% TOTAL');
    if (idx === -1) continue;
    const prefix = line.slice(0, idx).trim();
    const digits = prefix.split(' ').pop() ?? prefix;
    const value = Number.parseFloat(digits.trim());
    if (!Number.isNaN(value)) return Math.min(Math.max(value, 0), 100);
  }
  return undefined;
}

/** `/proc/meminfo`: MemTotal - MemAvailable (либо MemFree, если MemAvailable
 * недоступен — на старых ядрах его может не быть). */
export function parseMemInfo(output: string): { usedKB: number; totalKB: number } | undefined {
  let total: number | undefined;
  let available: number | undefined;
  let free: number | undefined;

  for (const line of output.split('\n')) {
    const parts = line.split(':');
    if (parts.length !== 2) continue;
    const key = parts[0].trim();
    const valueDigits = parts[1].trim().split(' ')[0];
    const value = Number.parseInt(valueDigits, 10);
    if (Number.isNaN(value)) continue;
    if (key === 'MemTotal') total = value;
    else if (key === 'MemAvailable') available = value;
    else if (key === 'MemFree') free = value;
  }

  if (total === undefined) return undefined;
  const used = total - (available ?? free ?? total);
  return { usedKB: Math.max(used, 0), totalKB: total };
}

export interface BatteryInfo {
  level?: number;
  temperature?: number;
  charging: boolean;
}

/** `dumpsys battery`: level/scale дают процент, temperature — в десятых
 * долях °C, заряд определяем по status==2 (BATTERY_STATUS_CHARGING) либо
 * по *_powered. */
export function parseBattery(output: string): BatteryInfo {
  let level: number | undefined;
  let scale: number | undefined;
  let temperature: number | undefined;
  let status: number | undefined;
  let anyPowered = false;

  const toInt = (s: string): number | undefined => {
    const n = Number.parseInt(s, 10);
    return Number.isNaN(n) ? undefined : n;
  };

  for (const rawLine of output.split('\n')) {
    const line = rawLine.trim();
    const colonIdx = line.indexOf(':');
    if (colonIdx === -1) continue;
    const key = line.slice(0, colonIdx).trim();
    const value = line.slice(colonIdx + 1).trim();
    switch (key) {
      case 'level':
        level = toInt(value);
        break;
      case 'scale':
        scale = toInt(value);
        break;
      case 'status':
        status = toInt(value);
        break;
      case 'temperature': {
        const n = Number.parseFloat(value);
        temperature = Number.isNaN(n) ? undefined : n / 10;
        break;
      }
      case 'AC powered':
      case 'USB powered':
      case 'Wireless powered':
      case 'Dock powered':
        if (value === 'true') anyPowered = true;
        break;
      default:
        break;
    }
  }

  const percent =
    level === undefined ? undefined : scale !== undefined && scale > 0 && scale !== 100 ? Math.round((level / scale) * 100) : level;
  const charging = status === 2 || anyPowered;
  return { level: percent, temperature, charging };
}

export function parseDeviceStats(
  cpuOutput: string,
  memOutput: string,
  batteryOutput: string,
  timestamp: number = Date.now()
): DeviceStats {
  const mem = parseMemInfo(memOutput);
  const battery = parseBattery(batteryOutput);
  return {
    cpuPercent: parseCpuPercent(cpuOutput),
    memUsedKB: mem?.usedKB ?? 0,
    memTotalKB: mem?.totalKB ?? 0,
    batteryLevel: battery.level,
    batteryTemperature: battery.temperature,
    isCharging: battery.charging,
    timestamp,
  };
}
